CIP-XXXX/README.md

```diff
@@ -0,0 +1,84 @@
+---
+CIP: XXXX
+Title: Arbitrary Script as Native Script spending conditions
+Authors: Sebastien Guillemot <seba@dcspark.io>
+Comments-URI: TBD
+Status: Draft
+Type: Standards
+Created: 2022-07-27
+License: CC-BY-4.0
+---
+
+# Abstract
+
+Native scripts are often easier to work with as any application that allows users to enter an ADA address to receive funds supports native scripts (the same is not true for Plutus scripts as the app would need to know how to structure the datum). However, limited composability between native scripts and Plutus scripts limits leveraging this fact.
+
+This CIP introduces a way to use native scripts as a starting point for more complex interactions which helps unlock use cases such as simple proxy contracts.
+
+# Motivation
+
+Suppose that you are part of a DAO whose funds are managed by a Plutus contract. Your DAO, in order to receive payments, would like receive ADA or tokens to its script address directly. However, this is non-trivial because applications cannot sent to arbitrary Plutus scripts as they do not know how to structure the datum or what other kind of restrictions may exist for this contract.
```

Comment:
Should this not be possible today, this would likely need to go into a `PlutusV3`, but changing the native scripts syntax is about the same size of change (IIRC there is even `SimpleScriptV2` as a language and this would then be `SimpleScriptV3`).

Reply:
> Are you sure that a UTxO with DatumNone paid to a plutus script is unspendable?
This is unfortunately the case. All spending scripts require datum (it is the 2-argument scripts that do not, such as for minting)

Reply:
> is about the same size of change

I disagree with this because any change to the Plutus context breaks same-transaction composability (i.e. you wouldn't be able to spend a PlutusV2 tx in the same tx as a PlutusV3 tx)

Reply:
> you cannot tell if a script hash inside of a transaction output is a native script or a plutus script without having the script.

@JaredCorduan Does that mean that the ledger could (or already does) only decide whether a script is `PlutusV1` or `PlutusV2` when being provided with a matching script as witness? Does that mean that it would then also decide which kind of context to prepare and pass to the plutus interpreter?

> any change to the Plutus context breaks same-transaction composability (i.e. you wouldn't be able to spend a PlutusV2 tx in the same tx as a PlutusV3 tx)

@SebastienGllmt If the above is true, then this could (should?) actually be possible.

Reply:
@ch1bo you are exactly right. in the ledger state, there is not way to tell if a script hash is a timelock script, a Plutus V1 script, or a Plutus V2 script until the script witness is revealed. indeed this means you cannot guess what kind of context to prepare until seeing the script.

Reply:
So regarding the need to create the datum with the right format for a dao, why don't the DAO owners/creators create an inline datum? It will be trivial for wallets to include as they won't have to construct the datum themselves.

Reply:
I just watched the video, it sounds like the problem @Jimbo4350 is that it is hard to communicate out-of-band about which datum to use. So I would imagine that maybe specifying a given output in the UTxO would be equally hard.

Also note that we do not support "reference datums". By this I mean that you cannot use a reference input which is connected to an inline datum and have that datum be supplied wherever it is needed inside the transaction (ie the way reference scripts work). Which means that using inline datums would be awkward due to them being consumed. (I am personally found of the "reference datum" idea).

Reply:
I'm a little confused as to the motivation also. You're communicating with your users already, by asking them to send funds to a particular address. Why can't you also tell them what datum to send? Is the issue that it just isn't *convenient*? Maybe we need an "extended address format" that lets you specify some datum bytes as well? What's the user journey that we're trying to support here?

Reply:
In my view, more automation is the primary goal.

Reply:
@JaredCorduan 
> Also note that we do not support "reference datums". By this I mean that you cannot use a reference input which is connected to an inline datum and have that datum be supplied wherever it is needed inside the transaction (ie the way reference scripts work). Which means that using inline datums would be awkward due to them being consumed. (I am personally found of the "reference datum" idea).

Can you elaborate on this? By this do you mean that when specifying a reference input, the datum still needs to be provided in the witness field of the tx, even if the input you're pointing to is an online datum?  

@michaelpj 
> Why can't you also tell them what datum to send?

Because wallets don't have a way to specify the datum in hand-built transactions; this could be added, for sure, but then asking the user to copy paste a cbor datum is a pretty bad user experience and prone to user error; also, picture the user flows around things like ada handles: user enters the ada handle in the wallet, it resolves to an address, and funds are sent to that address.  If the handle is sitting in a smart contract (for example, being sold on an NFT marketplace), this locks the funds sent there.

There are also cases where people have seen an address in their wallet from interacting with a dApp, and thought they could send more funds to that address to continue to interact with the dApp, resulting in locked funds.

Encoding the datum in the address is an interesting idea! That makes addresses a bit longer, but then resolving the ada handle address could encode the target datum.

It might also be possible for a community-standards approach to arise, where there is a standardized datum that the ecosystem resolves is the "don't lock these funds forever please" escape hatch datum that wallets could include, but that seems hard to get consensus on, and doesn't help with the problem of lost funds on existing contracts.

I don't have strong opinions on this specific proposal, but I thought I would weigh in with some of the end-user perspective I've been exposed to.

Reply:
> Can you elaborate on this? By this do you mean that when specifying a reference input, the datum still needs to be provided in the witness field of the tx, even if the input you're pointing to is an online datum?

I think the answer is yes, but let me try to spell it out. An inline datum in a transaction output can _only_ be used for the script (hash) inside the given output. You cannot just use a reference input whose corresponding output has an inline datum, and have that datum be totted around to all the places it is needed (like how reference scripts work).

Reply:
> I think the answer is yes, but let me try to spell it out. An inline datum in a transaction output can only be used for the script (hash) inside the given output.

I'm also not sure what you mean by this. The [PlutusContext](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Contexts.hs#L80) contains all the [reference inputs](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Contexts.hs#L65) of the transaction. These [outputs](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Tx.hs#L83) all contain the [datum](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Tx.hs#L62) which may be available inlined.

Reply:
> I'm also not sure what you mean by this. The [PlutusContext](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Contexts.hs#L80) contains all the [reference inputs](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Contexts.hs#L65) of the transaction. These [outputs](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Tx.hs#L83) all contain the [datum](https://github.com/input-output-hk/plutus/blob/c8d4364d0e639fef4d5b93f7d6c0912d992b54f9/plutus-ledger-api/src/PlutusLedgerApi/V2/Tx.hs#L62) which may be available inlined.

Yes, it is true that inline datums attached to reference inputs end up in the script context, but there is a requirement that each datum hash must have its preimage in the transaction witnesses. What I am saying as that reference inputs (which point to inline datums) cannot substitute for this requirement.

Reply:
> Encoding the datum in the address is an interesting idea! That makes addresses a bit longer, but then resolving the ada handle address could encode the target datum.

I am thinking of something like how URLs work. We are quite familiar with the idea of passing additional information in URLs via e.g. query parameters. That gives you the "one link to click" UX but with the ability to include extra payload to be sent. You could imagine an address format like `address?datum=<cbor hex>`. Maybe this is bad UX for other reasons, I just wanted to float it as an idea.

> There are also cases where people have seen an address in their wallet from interacting with a dApp, and thought they could send more funds to that address to continue to interact with the dApp, resulting in locked funds.

This seems pretty bad. I would think that on the wallet side you want to be pretty careful with letting people send money to script addresses, but I'm not sure what the ideal UX is.

Reply:
Exactly; some wallets are starting to warn about that, but the main problem is that the wallets can't tell the difference between a native script and a Plutus script without seeing the script bytes somewhere on-chain.  It's totally safe to send funds to a multi-sig address without a datum.  They both start 0x70.  Hence, I think, #310